File: app/src/main/java/com/damhoe/scoresheetskat/score/adapter/in/ui/ScoreFragment.java
package com.damhoe.scoresheetskat.score.adapter.in.ui;

import androidx.core.content.res.ResourcesCompat;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.damhoe.scoresheetskat.MainActivity;
import com.damhoe.scoresheetskat.R;
import com.damhoe.scoresheetskat.databinding.FragmentScoreBinding;
import com.damhoe.scoresheetskat.score.domain.ScoreEvent;
import com.damhoe.scoresheetskat.score.domain.ScoreEventType;
import com.damhoe.scoresheetskat.score.domain.ScoreRequest;
import com.damhoe.scoresheetskat.score.domain.SkatResult;
import com.damhoe.scoresheetskat.score.domain.SkatScore;
import com.damhoe.scoresheetskat.score.domain.SkatScoreCommand;
import com.damhoe.scoresheetskat.score.domain.SkatSuit;
import com.damhoe.scoresheetskat.shared_ui.utils.InsetsManager;
import com.damhoe.scoresheetskat.shared_ui.utils.LayoutMargins;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.button.MaterialButtonToggleGroup;
import com.google.android.material.chip.ChipGroup;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

public class ScoreFragment extends Fragment {
    @Inject
    ScoreViewModelFactory viewModelFactory;
    private ScoreViewModel viewModel;
    private SharedScoreResponseViewModel sharedViewModel;
    private ScoreEventType eventType;
    private FragmentScoreBinding binding;
    private final Map<Integer, SkatSuit> buttonSkatSuitMap = new HashMap<>();
    private final Map<Integer, Runnable> buttonPlayerActionMap = new HashMap<>();
    private final Map<Integer, Runnable> playerPositionCheckButtonMap = new HashMap<>();
    private final Map<SkatSuit, Runnable> suitCheckButtonMap = new HashMap<>();

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        ((MainActivity) requireActivity()).appComponent.inject(this);
        super.onCreate(savedInstanceState);

        ScoreRequest scoreRequest = requireArguments().getParcelable("scoreRequest");
        if (scoreRequest == null) {
            throw new RuntimeException("Score request should have nonnull value.");
        }

        viewModelFactory.setScoreRequest(scoreRequest);
        eventType = scoreRequest.getScoreId() == -1L ? ScoreEventType.CREATE : ScoreEventType.UPDATE;
        viewModel = viewModelFactory.create(ScoreViewModel.class);

        sharedViewModel = new ViewModelProvider(requireActivity()).get(SharedScoreResponseViewModel.class);
        sharedViewModel.reset();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(getLayoutInflater(), R.layout.fragment_score, container, false);
        initializeButtonSkatSuitMap();
        initializeButtonPlayerActionMap();

        binding.toggleGroupSoloPlayer.addOnButtonCheckedListener((group, checkedId, isChecked) -> {
            if (isChecked) {
                handlePlayerButtonCheck(checkedId);
            }
        });
        binding.toggleGroupResult.addOnButtonCheckedListener((group, checkedId, isChecked) -> {
            if (isChecked) {
                if (checkedId == R.id.buttonOverbid) {
                    viewModel.setResult(SkatResult.OVERBID);
                } else if (checkedId == R.id.buttonWon){
                    viewModel.setResult(SkatResult.WON);
                } else {
                    viewModel.setResult(SkatResult.LOST);
                }
            }
        });
        binding.toggleGroupSuit.addOnButtonCheckedListener(((group, checkedId, isChecked) -> {
            if (isChecked) {
                binding.toggleGroupSpecialSuit.clearChecked();
                handleSuitButtonCheck(checkedId);
            }
        }));
        binding.toggleGroupSpecialSuit.addOnButtonCheckedListener(((group, checkedId, isChecked) -> {
            if (isChecked) {
                binding.toggleGroupSuit.clearChecked();
                handleSuitButtonCheck(checkedId);
            }
        }));
        binding.announcementsChips.setOnCheckedStateChangeListener((group, checkedIds) -> {
            binding.content.requestLayout();
            handleAnnouncementChips(checkedIds);
        });
        binding.winLevelChips.setOnCheckedStateChangeListener(((group, checkedIds) -> {
            binding.content.requestLayout();
            handleWinLevelChips(checkedIds);
        }));
        binding.spitzenSlider.addOnChangeListener((slider, value, fromUser) -> {
            binding.spitzenLabel.setText(String.valueOf((int) value));
            viewModel.setSpitzen((int) value);
        });
        binding.spitzenRemoveButton.setOnClickListener(
                view -> binding.spitzenSlider.setValue(binding.spitzenSlider.getValue() - 1)
        );
        binding.spitzenAddButton.setOnClickListener(
                view -> binding.spitzenSlider.setValue(binding.spitzenSlider.getValue() + 1)
        );
        binding.scoreDoneButton.setOnClickListener(view -> {
            saveScore();
            findNavController().navigateUp();
        });
        return binding.getRoot();
    }

    private void handleAnnouncementChips(@NonNull List<Integer> checkedIds) {
        // Schwarz is checked only if Schneider is also checked
        if (checkedIds.contains(R.id.schwarz_announced_chip) &&
                !checkedIds.contains(R.id.schneider_announced_chip)) {
            binding.announcementsChips.check(R.id.schneider_announced_chip);
            return;
        }

        // Set viewModel data
        viewModel.setSchneiderAnnounced(checkedIds.contains(R.id.schneider_announced_chip));
        viewModel.setSchwarzAnnounced(checkedIds.contains(R.id.schwarz_announced_chip));

    }

    private void handleWinLevelChips(@NonNull List<Integer> checkedIds) {
        // Schwarz is checked only if Schneider is also checked
        if (checkedIds.contains(R.id.schwarz_chip) &&
                !checkedIds.contains(R.id.schneider_chip)) {
            binding.winLevelChips.check(R.id.schneider_chip);
            return;
        }
        viewModel.setHand(checkedIds.contains(R.id.hand_chip));
        viewModel.setOuvert(checkedIds.contains(R.id.ouvert_chip));
        viewModel.setSchneider(checkedIds.contains(R.id.schneider_chip));
        viewModel.setSchwarz(checkedIds.contains(R.id.schwarz_chip));
    }

    private NavController findNavController() {
        return Navigation.findNavController(requireActivity(), R.id.nav_host_fragment);
    }

    private void handlePlayerButtonCheck(int buttonId) {
        Runnable action = buttonPlayerActionMap.get(buttonId);
        if (action != null) {
            action.run();
        }
    }

    private void handleSuitButtonCheck(int buttonId) {
        viewModel.setSuit(buttonSkatSuitMap.get(buttonId));
    }

    // Action if the button gets checked
    private void initializeButtonPlayerActionMap() {
        buttonPlayerActionMap.put(binding.buttonPasse.getId(), () -> viewModel.setPasse());
        buttonPlayerActionMap.put(binding.buttonPlayer1.getId(),
                () -> viewModel.setPlayerPosition(viewModel.getPlayerPositions()[0]));
        buttonPlayerActionMap.put(binding.buttonPlayer2.getId(),
                () -> viewModel.setPlayerPosition(viewModel.getPlayerPositions()[1]));
        buttonPlayerActionMap.put(binding.buttonPlayer3.getId(),
                () -> viewModel.setPlayerPosition(viewModel.getPlayerPositions()[2]));

        playerPositionCheckButtonMap.put(viewModel.getPlayerPositions()[0], () -> binding.toggleGroupSoloPlayer.check(binding.buttonPlayer1.getId()));
        playerPositionCheckButtonMap.put(viewModel.getPlayerPositions()[1], () -> binding.toggleGroupSoloPlayer.check(binding.buttonPlayer2.getId()));
        playerPositionCheckButtonMap.put(viewModel.getPlayerPositions()[2], () -> binding.toggleGroupSoloPlayer.check(binding.buttonPlayer3.getId()));
    }

    private void initializeButtonSkatSuitMap() {
        buttonSkatSuitMap.put(binding.buttonClubs.getId(), SkatSuit.CLUBS);
        buttonSkatSuitMap.put(binding.buttonSpades.getId(), SkatSuit.SPADES);
        buttonSkatSuitMap.put(binding.buttonHearts.getId(), SkatSuit.HEARTS);
        buttonSkatSuitMap.put(binding.buttonDiamonds.getId(), SkatSuit.DIAMONDS);
        buttonSkatSuitMap.put(binding.buttonGrand.getId(), SkatSuit.GRAND);
        buttonSkatSuitMap.put(binding.buttonNull.getId(), SkatSuit.NULL);

        suitCheckButtonMap.put(SkatSuit.CLUBS, () -> binding.toggleGroupSuit.check(binding.buttonClubs.getId()));
        suitCheckButtonMap.put(SkatSuit.SPADES, () -> binding.toggleGroupSuit.check(binding.buttonSpades.getId()));
        suitCheckButtonMap.put(SkatSuit.HEARTS, () -> binding.toggleGroupSuit.check(binding.buttonHearts.getId()));
        suitCheckButtonMap.put(SkatSuit.DIAMONDS, () -> binding.toggleGroupSuit.check(binding.buttonDiamonds.getId()));
        suitCheckButtonMap.put(SkatSuit.GRAND, () -> binding.toggleGroupSpecialSuit.check(binding.buttonGrand.getId()));
        suitCheckButtonMap.put(SkatSuit.NULL, () -> binding.toggleGroupSpecialSuit.check(binding.buttonNull.getId()));
    }

    private void showOrHideCheckedIcon(MaterialButtonToggleGroup group, int checkedId, boolean isChecked) {
        if (isChecked) {
            ((MaterialButton) group.findViewById(checkedId)).setIcon(
                    ResourcesCompat.getDrawable(getResources(),
                            R.drawable.ic_done_black_24dp, null));
        }
        else {
            ((MaterialButton) group.findViewById(checkedId)).setIcon(null);
        }
    }

    /** @noinspection DataFlowIssue*/
    private void initializeUI() {
        String[] names = viewModel.getPlayerNames();
        if (names.length != 3) {
            throw new RuntimeException("Expected 3 player names" +
                    " in observed LiveData but got " + names.length);
        }
        binding.buttonPlayer1.setText(names[0]);
        binding.buttonPlayer2.setText(names[1]);
        binding.buttonPlayer3.setText(names[2]);

        SkatScoreCommand command = viewModel.getScoreCommand().getValue();
        if (command != null) {
            if (command.isPasse()) {
                binding.toggleGroupSoloPlayer.check(R.id.buttonPasse);
                binding.spitzenSlider.setValue(1);
                return;
            }

            if (command.isLost()) {
                binding.toggleGroupResult.check(R.id.buttonLost);
            } else if (command.isWon()) {
                binding.toggleGroupResult.check(R.id.buttonWon);
            } else {
                binding.toggleGroupResult.check(R.id.buttonOverbid);
            }

            playerPositionCheckButtonMap.get(command.getPlayerPosition()).run();
            suitCheckButtonMap.get(command.getSuit()).run();
            binding.spitzenSlider.setValue(command.getSpitzen());
            binding.handChip.setChecked(command.isHand());
            binding.ouvertChip.setChecked(command.isOuvert());
            binding.schneiderChip.setChecked(command.isSchneider());
            binding.schwarzChip.setChecked(command.isSchwarz());
            binding.schneiderAnnouncedChip.setChecked(command.isSchneiderAnnounced());
            binding.schwarzAnnouncedChip.setChecked(command.isSchwarzAnnounced());
            binding.spitzenSlider.setValue(command.getSpitzen());
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        // Setup navigation
        NavController navController = findNavController();
        AppBarConfiguration appBarConfiguration =
                ((MainActivity)requireActivity()).getAppBarConfiguration();
        NavigationUI.setupWithNavController(binding.toolbar, navController, appBarConfiguration);

        // Set insets
        InsetsManager.applyStatusBarInsets(binding.appbarLayout);
        int marginRight = getResources().getDimensionPixelSize(R.dimen.fab_margin_right);
        int marginBottom = getResources().getDimensionPixelSize(R.dimen.fab_margin_bottom);
        LayoutMargins defaultMargins =
                new LayoutMargins(0, 0, marginRight, marginBottom);
        InsetsManager.applyNavigationBarInsets(binding.scoreDoneButton, defaultMargins);
        InsetsManager.applyNavigationBarInsets(binding.content);

        viewModel.isSuitsEnabled.observe(getViewLifecycleOwner(), this::enableSuitButtons);
        viewModel.isHandEnabled.observe(getViewLifecycleOwner(), isEnabled -> {
            binding.handChip.setEnabled(isEnabled);
        });
        viewModel.isSchneiderSchwarzEnabled.observe(getViewLifecycleOwner(),
                this::enableSchneiderSchwarzChips);
        viewModel.isOuvertEnabled.observe(getViewLifecycleOwner(),
                isEnabled -> binding.ouvertChip.setEnabled(isEnabled)
        );
        viewModel.isSpitzenEnabled.observe(getViewLifecycleOwner(),
                isEnabled -> binding.spitzenSlider.setEnabled(isEnabled)
        );
        viewModel.isIncreaseSpitzenEnabled.observe(getViewLifecycleOwner(),
                isEnabled -> binding.spitzenAddButton.setEnabled(isEnabled)
        );
        viewModel.isDecreaseSpitzenEnabled.observe(getViewLifecycleOwner(),
                isEnabled -> binding.spitzenRemoveButton.setEnabled(isEnabled)
        );
        viewModel.isResultsEnabled.observe(getViewLifecycleOwner(), isEnabled -> {
            binding.buttonWon.setEnabled(isEnabled);
            binding.buttonLost.setEnabled(isEnabled);
            binding.buttonOverbid.setEnabled(isEnabled);
        });
        viewModel.getSkatResult.observe(getViewLifecycleOwner(), skatResult -> {
            if (skatResult == SkatResult.WON) {
                binding.toggleGroupResult.check(R.id.buttonWon);
            } else if (skatResult == SkatResult.LOST) {
                binding.toggleGroupResult.check(R.id.buttonLost);
            } else if (skatResult == SkatResult.OVERBID) {
                binding.toggleGroupResult.check(R.id.buttonOverbid);
            } else {
                binding.toggleGroupResult.clearChecked();
            }
        });
        viewModel.getSuit.observe(getViewLifecycleOwner(), skatSuit -> {
            if (skatSuit == SkatSuit.INVALID) {
                binding.announcementsChips.clearCheck();
                binding.winLevelChips.clearCheck();
                binding.spitzenSlider.setValue(binding.spitzenSlider.getValueFrom());
                binding.toggleGroupSuit.clearChecked();
            } else if (skatSuit == SkatSuit.CLUBS) {
                binding.toggleGroupSuit.check(R.id.buttonClubs);
            } else if (skatSuit == SkatSuit.SPADES) {
                binding.toggleGroupSuit.check(R.id.buttonSpades);
            } else if (skatSuit == SkatSuit.HEARTS) {
                binding.toggleGroupSuit.check(R.id.buttonHearts);
            } else if (skatSuit == SkatSuit.DIAMONDS) {
                binding.toggleGroupSuit.check(R.id.buttonDiamonds);
            } else if (skatSuit == SkatSuit.NULL) {
                binding.toggleGroupSuit.check(R.id.buttonNull);
            } else if (skatSuit == SkatSuit.GRAND) {
                binding.toggleGroupSuit.check(R.id.buttonGrand);
            }
        });

        initializeUI();
    }

    private void enableSchneiderSchwarzChips(Boolean isEnabled) {
        binding.schneiderChip.setEnabled(isEnabled);
        binding.schwarzChip.setEnabled(isEnabled);
        binding.schneiderAnnouncedChip.setEnabled(isEnabled);
        binding.schwarzAnnouncedChip.setEnabled(isEnabled);
    }

    private void enableSuitButtons(Boolean isEnabled) {
        binding.buttonClubs.setEnabled(isEnabled);
        binding.buttonHearts.setEnabled(isEnabled);
        binding.buttonDiamonds.setEnabled(isEnabled);
        binding.buttonSpades.setEnabled(isEnabled);
        binding.buttonNull.setEnabled(isEnabled);
        binding.buttonGrand.setEnabled(isEnabled);
        binding.buttonHearts.setAlpha(isEnabled ? 1f : 0.38f);
        binding.buttonDiamonds.setAlpha(isEnabled ? 1f : 0.38f);
    }

    public void saveScore() {
        SkatScore score;
        switch (eventType) {
            case CREATE:
                score = viewModel.createScore();
                break;
            case UPDATE:
                score = viewModel.updateScore();
                break;
            default:
                score = null;
        }
        if (score != null) {
            sharedViewModel.setScoreEvent(new ScoreEvent(score, eventType));
        }
    }
}

File: app/src/main/java/com/damhoe/skatscores/game/score/adapter/in/ui/ScoreFragment.kt
package com.damhoe.skatscores.game.score.adapter.`in`.ui

import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.NavController
import androidx.navigation.Navigation
import androidx.navigation.ui.NavigationUI.setupWithNavController
import com.damhoe.skatscores.R
import com.damhoe.skatscores.databinding.FragmentScoreBinding
import com.damhoe.skatscores.game.adapter.presentation.GameActivity
import com.damhoe.skatscores.game.adapter.presentation.GameViewModelFactory
import com.damhoe.skatscores.game.adapter.presentation.SkatGameViewModel
import com.damhoe.skatscores.game.domain.score.ScoreRequest
import com.damhoe.skatscores.game.domain.score.SkatOutcome.LOST
import com.damhoe.skatscores.game.domain.score.SkatOutcome.OVERBID
import com.damhoe.skatscores.game.domain.score.SkatOutcome.PASSE
import com.damhoe.skatscores.game.domain.score.SkatOutcome.WON
import com.damhoe.skatscores.game.domain.score.SkatSuit.CLUBS
import com.damhoe.skatscores.game.domain.score.SkatSuit.DIAMONDS
import com.damhoe.skatscores.game.domain.score.SkatSuit.GRAND
import com.damhoe.skatscores.game.domain.score.SkatSuit.HEARTS
import com.damhoe.skatscores.game.domain.score.SkatSuit.INVALID
import com.damhoe.skatscores.game.domain.score.SkatSuit.NULL
import com.damhoe.skatscores.game.domain.score.SkatSuit.SPADES
import com.damhoe.skatscores.shared_ui.utils.InsetsManager
import com.damhoe.skatscores.shared_ui.utils.LayoutMargins
import javax.inject.Inject

class ScoreFragment : Fragment()
{
    /**
     * Score fragment is responsible for the creation and update process of scores.
     *
     * Two view model instances are used to manage the data:
     *      (i) Score view model manages the current score data that is
     *          subject to user interaction
     *      (ii) Game view model lives at the activity level and holds the current game
     *          It is responsible for persisting the score if accepted
     */
    companion object
    {
        const val SCORE_REQUEST_KEY: String = "ScoreRequest"
    }

    private val maxSpitzenValueTo = 11f
    private val normalSpitzenValueTo = 4f

    private val scoreRequest: ScoreRequest by lazy {
        arguments?.let {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
            {
                it.getParcelable(SCORE_REQUEST_KEY, ScoreRequest::class.java)
            } else
            {
                it.getParcelable(SCORE_REQUEST_KEY)
            }
        } ?: throw IllegalArgumentException("Score fragment is missing score request.")
    }

    @Inject
    lateinit var scoreViewModelFactory: ScoreViewModelFactory
    private val scoreViewModel: ScoreViewModel by viewModels {
        scoreViewModelFactory.also { it.scoreRequest = scoreRequest }
    }

    // Game view model lives at activity scope
    @Inject
    lateinit var gameViewModelFactory: GameViewModelFactory
    private val gameViewModel: SkatGameViewModel
            by viewModels({ requireActivity() }) { gameViewModelFactory }

    private lateinit var binding: FragmentScoreBinding

    override fun onCreate(savedInstanceState: Bundle?)
    {
        (requireActivity() as GameActivity).appComponent.inject(this)
        super.onCreate(savedInstanceState)
    }

    private fun addListenersForToggleButtons() = binding.apply {
        toggleGroupSoloPlayer.addOnButtonCheckedListener { _, checkedId, isChecked ->
            if (isChecked)
            {
                handlePlayerButtonCheck(checkedId)
            }
        }

        toggleGroupResult.addOnButtonCheckedListener { _, checkedId, isChecked ->
            if (isChecked)
            {
                handleResultButtonCheck(checkedId)
            }
        }

        toggleGroupSuit.addOnButtonCheckedListener { _, checkedId, isChecked ->
            if (isChecked)
            {
                // Also clear special suits
                toggleGroupSpecialSuit.clearChecked()
                handleSuitButtonCheck(checkedId)
            }
        }

        toggleGroupSpecialSuit.addOnButtonCheckedListener { _, checkedId, isChecked ->
            if (isChecked)
            {
                // Also clear special suits
                toggleGroupSuit.clearChecked()
                handleSuitButtonCheck(checkedId)
            }
        }
    }

    private fun handleResultButtonCheck(checkedId: Int)
    {
        when (checkedId)
        {
            R.id.buttonOverbid -> scoreViewModel.setResult(OVERBID)
            R.id.buttonWon -> scoreViewModel.setResult(WON)
            R.id.buttonLost -> scoreViewModel.setResult(LOST)
        }
    }

    private fun addListenersForChips() = binding.apply {
        announcementsChips.setOnCheckedStateChangeListener { _, checkedIds: List<Int> ->
            content.requestLayout()
            handleAnnouncementChips(checkedIds)
        }

        winLevelChips.setOnCheckedStateChangeListener { _, checkedIds: List<Int> ->
            content.requestLayout()
            handleWinLevelChips(checkedIds)
        }
    }

    private fun updateSpitzenButtons(spitzen: Float)
    {
        with(binding) {
            spitzenAddButton.isEnabled = spitzen < spitzenSlider.valueTo
            spitzenRemoveButton.isEnabled = spitzen > spitzenSlider.valueFrom
        }
    }

    private fun addListenersForSpitzen() = binding.apply {
        spitzenSlider.addOnChangeListener { _, value, _ ->
            spitzenLabel.text = String.format("%.0f", value)
            scoreViewModel.setSpitzen(value.toInt())
            updateSpitzenButtons(value)
        }
        spitzenRemoveButton.setOnClickListener { spitzenSlider.value -= 1 }
        spitzenAddButton.setOnClickListener { spitzenSlider.value += 1 }

        spitzenSwitch.setOnCheckedChangeListener { _, isChecked ->
            spitzenSlider.valueTo = if (isChecked)
            {
                maxSpitzenValueTo
            } else
            {
                if (spitzenSlider.value > normalSpitzenValueTo)
                {
                    spitzenSlider.value = normalSpitzenValueTo
                }
                normalSpitzenValueTo
            }
            updateSpitzenButtons(spitzenSlider.value)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View
    {
        binding = DataBindingUtil.inflate(
            layoutInflater,
            R.layout.fragment_score, container, false
        )

        addListenersForToggleButtons()
        addListenersForChips()
        addListenersForSpitzen()

        binding.scoreDoneButton.setOnClickListener {
            saveScore()
            findNavController().navigateUp()
        }

        return binding.root
    }

    private fun handleAnnouncementChips(checkedIds: List<Int>)
    {
        // Schwarz is checked only if Schneider is also checked
        if (checkedIds.contains(R.id.schwarz_announced_chip) &&
            !checkedIds.contains(R.id.schneider_announced_chip)
        )
        {
            binding.announcementsChips.check(R.id.schneider_announced_chip)
            return
        }

        // Set viewModel data
        scoreViewModel.setSchneiderAnnounced(checkedIds.contains(R.id.schneider_announced_chip))
        scoreViewModel.setSchwarzAnnounced(checkedIds.contains(R.id.schwarz_announced_chip))
    }

    private fun handleWinLevelChips(checkedIds: List<Int>)
    {
        // Schwarz is checked only if Schneider is also checked
        if (checkedIds.contains(R.id.schwarz_chip) &&
            !checkedIds.contains(R.id.schneider_chip)
        )
        {
            binding.winLevelChips.check(R.id.schneider_chip)
            return
        }
        scoreViewModel.setHand(checkedIds.contains(R.id.hand_chip))
        scoreViewModel.setOuvert(checkedIds.contains(R.id.ouvert_chip))
        scoreViewModel.setSchneider(checkedIds.contains(R.id.schneider_chip))
        scoreViewModel.setSchwarz(checkedIds.contains(R.id.schwarz_chip))
    }

    private fun findNavController() =
        Navigation.findNavController(requireActivity(), R.id.nav_host_fragment)

    private fun handlePlayerButtonCheck(buttonId: Int) = with(binding) {
        scoreViewModel.apply {
            when (buttonId)
            {
                buttonPasse.id -> setPasse()
                buttonPlayer1.id -> selectPlayer(position = 0)
                buttonPlayer2.id -> selectPlayer(position = 1)
                buttonPlayer3.id -> selectPlayer(position = 2)
            }
        }
    }

    private fun handleSuitButtonCheck(buttonId: Int) = with(binding) {
        scoreViewModel.apply {
            when (buttonId)
            {
                buttonClubs.id -> setSuit(CLUBS)
                buttonSpades.id -> setSuit(SPADES)
                buttonHearts.id -> setSuit(HEARTS)
                buttonDiamonds.id -> setSuit(DIAMONDS)
                buttonGrand.id -> setSuit(GRAND)
                buttonNull.id -> setSuit(NULL)
            }
        }
    }

    private fun writePlayerNames() = binding.also {
        with(scoreViewModel.playerNames) {
            it.buttonPlayer1.text = this[0]
            it.buttonPlayer2.text = this[1]
            it.buttonPlayer3.text = this[2]
        }
    }

    private fun checkPlayer(position: Int) = binding.apply {
        when (position)
        {
            0 -> buttonPlayer1.isChecked = true
            1 -> buttonPlayer2.isChecked = true
            2 -> buttonPlayer3.isChecked = true
        }
    }

    /** @noinspection DataFlowIssue
     */
    private fun initializeUI()
    {
        writePlayerNames()

        val positionMap = scoreViewModel.playerPositions

        scoreViewModel.scoreCommand.value?.let { cmd ->
            binding.apply {

                if (cmd.isPasse)
                {
                    toggleGroupSoloPlayer.check(R.id.buttonPasse)
                    return
                }

                val playerPosition = positionMap[cmd.playerPosition]
                checkPlayer(playerPosition)

                toggleGroupResult.apply {
                    when
                    {
                        cmd.isWon -> check(R.id.buttonWon)
                        cmd.isLost -> check(R.id.buttonLost)
                        cmd.isOverbid -> check(R.id.buttonOverbid)
                    }
                }

                // Init slider
                spitzenSwitch.isChecked = cmd.spitzen > normalSpitzenValueTo
                spitzenSlider.valueTo = if (cmd.spitzen > normalSpitzenValueTo)
                {
                    maxSpitzenValueTo
                } else
                {
                    normalSpitzenValueTo
                }
                spitzenSlider.value = cmd.spitzen.toFloat()

                // Init win level chips
                handChip.isChecked = cmd.isHand
                ouvertChip.isChecked = cmd.isOuvert
                schneiderChip.isChecked = cmd.isSchneider
                schwarzChip.isChecked = cmd.isSchwarz
                schneiderAnnouncedChip.isChecked = cmd.isSchneiderAnnounced
                schwarzAnnouncedChip.isChecked = cmd.isSchwarzAnnounced
            }
        }
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?)
    {
        super.onViewCreated(view, savedInstanceState)

        // Setup navigation
        val navController: NavController = findNavController()
        setupWithNavController(binding.toolbar, navController)

        // Set insets
        InsetsManager.applyStatusBarInsets(binding.appbarLayout)
        val marginRight = resources.getDimensionPixelSize(R.dimen.fab_margin_right)
        val marginBottom = resources.getDimensionPixelSize(R.dimen.fab_margin_bottom)
        val defaultMargins = LayoutMargins(0, 0, marginRight, marginBottom)
        InsetsManager.applyNavigationBarInsets(binding.scoreDoneButton, defaultMargins)
        InsetsManager.applyNavigationBarInsets(binding.content)

        with(scoreViewModel) {
            isSuitsEnabled
                .observe(viewLifecycleOwner) { changeSuitButtonAccessibility(it) }
            isHandEnabled
                .observe(viewLifecycleOwner) { binding.handChip.isEnabled = it }
            isSchneiderSchwarzEnabled
                .observe(viewLifecycleOwner) { changeAnnouncementsChipsAccessibility(it) }
            isOuvertEnabled
                .observe(viewLifecycleOwner) { binding.ouvertChip.isEnabled = it }
            isSpitzenEnabled
                .observe(viewLifecycleOwner) {
                    binding.spitzenSlider.isEnabled = it
                    binding.spitzenSwitch.isEnabled = it
                }

            isResultsEnabled.observe(viewLifecycleOwner) { b ->
                binding.apply {
                    listOf(buttonWon, buttonLost, buttonOverbid).forEach { it.isEnabled = b }
                }
            }

            skatResult.observe(viewLifecycleOwner) { outcome ->
                outcome?.let {
                    when (it)
                    {
                        WON -> binding.toggleGroupResult.check(R.id.buttonWon)
                        LOST -> binding.toggleGroupResult.check(R.id.buttonLost)
                        OVERBID -> binding.toggleGroupResult.check(R.id.buttonOverbid)
                        PASSE -> binding.toggleGroupResult.clearChecked()
                    }
                }
            }

            suit.observe(viewLifecycleOwner) { skatSuit ->
                skatSuit?.let {
                    when (it)
                    {
                        INVALID -> binding.apply {
                            announcementsChips.clearCheck()
                            winLevelChips.clearCheck()
                            spitzenSlider.value = spitzenSlider.valueFrom
                            toggleGroupSuit.clearChecked()
                        }

                        CLUBS -> binding.toggleGroupSuit.check(R.id.buttonClubs)
                        SPADES -> binding.toggleGroupSuit.check(R.id.buttonSpades)
                        HEARTS -> binding.toggleGroupSuit.check(R.id.buttonHearts)
                        DIAMONDS -> binding.toggleGroupSuit.check(R.id.buttonDiamonds)
                        NULL -> binding.toggleGroupSuit.check(R.id.buttonNull)
                        GRAND -> binding.toggleGroupSuit.check(R.id.buttonGrand)
                    }
                }
            }
        }

        initializeUI()
    }

    private fun changeAnnouncementsChipsAccessibility(isEnabled: Boolean) = with(binding) {
        listOf(schneiderChip, schwarzChip, schneiderAnnouncedChip, schwarzAnnouncedChip)
            .forEach { it.isEnabled = isEnabled }
    }

    private fun changeSuitButtonAccessibility(isEnabled: Boolean) = with(binding) {
        listOf(buttonClubs, buttonSpades, buttonHearts, buttonDiamonds, buttonNull, buttonGrand)
            .forEach { it.isEnabled = isEnabled }
    }

    private fun saveScore()
    {
        when (scoreRequest)
        {
            is ScoreRequest.CreateScoreRequest ->
            {
                val score = scoreViewModel.createScore()
                gameViewModel.addScore(score)
            }

            is ScoreRequest.UpdateScoreRequest ->
            {
                val score = scoreViewModel.updateScore()
                gameViewModel.updateScore(score)
            }
        }
    }
}